### PR TITLE
Add amp-fx="fade-in" - scroll triggered timed animation. 

### PR DESCRIPTION
**Pull request by nainar**
This PR fixes #14351 adds the following functionality:
* adds the ability to accept `amp-fx="fade-in"` which triggers a timed fade in animation 
* customizing the fade in animation by specifying:
  * duration of animation
  * easing curve of the animation
  * margin at which the animation should be triggered

### CODE REVIEW

File: examples/article-fade-in.amp.html [hunk 1]
```diff
@@ -0,0 +1,112 @@
+<!doctype html>
```
Comment (nainar) [inline examples/article-fade-in.amp.html]: New compared to #14276

File: extensions/amp-fx-collection/0.1/amp-fx-collection.js [hunk 2]
```diff
@@ -134,6 +128,10 @@ class AmpFxCollection {
 
     // Validate that we support the requested fx types.
     fxTypes.forEach(fxType => {
+      if (fxType == FxType.FADE_IN) {
```
Comment (nainar) [inline extensions/amp-fx-collection/0.1/amp-fx-collection.js]: New compared to #14276

File: extensions/amp-fx-collection/0.1/providers/amp-fx-presets-utils.js [hunk 3]
```diff
@@ -0,0 +1,47 @@
+/**
```
Comment (nainar) [inline extensions/amp-fx-collection/0.1/providers/amp-fx-presets-utils.js]: New compared to #14276

File: extensions/amp-fx-collection/0.1/providers/amp-fx-presets.js [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+              fxElement.setIsMutateScheduled(false);
+              // Translate the element offset pixels.
+              setStyles(fxElement.getElement(),
+                  {transform:
+                    `translateY(${fxElement.getOffset().toFixed(0)}px)`});
+            });
+      }
+    },
+  },
+  'fade-in': {
```
Comment (nainar) [inline extensions/amp-fx-collection/0.1/providers/amp-fx-presets.js]: New compared to #14276

File: extensions/amp-fx-collection/0.1/providers/fx-provider.js [hunk 5]
```diff
@@ -17,27 +17,35 @@
 import {
   PositionObserverFidelity,
 } from '../../../../src/service/position-observer/position-observer-worker';
+import {Presets} from './amp-fx-presets';
 import {Services} from '../../../../src/services';
-import {dev, user} from '../../../../src/log';
+import {convertEasingKeyword} from './amp-fx-presets-utils';
```
Comment (nainar) [inline extensions/amp-fx-collection/0.1/providers/fx-provider.js]: New compared to #14276

File: extensions/amp-fx-collection/0.1/providers/fx-provider.js [hunk 6]
```diff
@@ -17,27 +17,35 @@
...
 import {getServiceForDoc} from '../../../../src/service';
 import {
   installPositionObserverServiceForDoc,
 } from '../../../../src/service/position-observer/position-observer-impl';
-import {setStyle, setStyles} from '../../../../src/style';
+import {setStyles} from '../../../../src/style';
 
-const FACTOR_ATTR = 'data-parallax-factor';
+const installStyles = {
```
Comment (nainar) [inline extensions/amp-fx-collection/0.1/providers/fx-provider.js]: New compared to #14276

File: extensions/amp-fx-collection/0.1/providers/fx-provider.js [hunk 7]
```diff
@@ -87,36 +99,46 @@ class ParallaxElement {
...
+
+    /** @private @string */
+    this.fxType_ = fxType;
+
+    Presets[this.fxType_].userAsserts(element);
+
+    /** @private {number} */
+    this.factor_ = parseFloat(element.getAttribute('data-parallax-factor'));
+
+    /** @private {number} */
```
Comment (nainar) [inline extensions/amp-fx-collection/0.1/providers/fx-provider.js]: New compared to #14276

File: extensions/amp-fx-collection/0.1/test/validator-amp-fx-collection.html [hunk 8]
```diff
@@ -33,5 +33,10 @@
   <div>
     <h1 amp-fx="parallax"></h1>
   </div>
+  <div amp-fx="fade-in"></div>
```
Comment (nainar) [inline extensions/amp-fx-collection/0.1/test/validator-amp-fx-collection.html]: New compared to #14276

File: extensions/amp-fx-collection/0.1/test/validator-amp-fx-collection.out [hunk 9]
```diff
@@ -34,5 +34,10 @@ PASS
 |    <div>
 |      <h1 amp-fx="parallax"></h1>
 |    </div>
+|    <div amp-fx="fade-in"></div>
```
Comment (nainar) [inline extensions/amp-fx-collection/0.1/test/validator-amp-fx-collection.out]: New compared to #14276

File: extensions/amp-fx-collection/amp-fx-collection.md [hunk 10]
```diff
@@ -79,6 +79,38 @@ In this example, as the user scrolls the page, the h1 element scrolls faster rel
 </h1>
 ```
 
+### fade-in
```
Comment (nainar) [inline extensions/amp-fx-collection/amp-fx-collection.md]: New compared to #14276

File: tools/experiments/experiments.js [hunk 11]
```diff
@@ -288,6 +288,12 @@ const EXPERIMENTS = [
     spec: 'https://github.com/ampproject/amphtml/issues/9074',
     cleanupissue: 'https://github.com/ampproject/amphtml/issues/14263',
   },
+  {
```
Comment (nainar) [inline tools/experiments/experiments.js]: New compared to #14276

File: validator/validator-main.protoascii [hunk 12]
```diff
@@ -4524,7 +4524,7 @@ attr_lists: {
   # https://www.ampproject.org/docs/reference/components/amp-fx-collection
   attrs: {
     name: "amp-fx"
-    value_casei: "parallax"
+    value_regex_casei: "(fade-in|parallax)"
```
Comment (nainar) [inline validator/validator-main.protoascii]: New compared to #14276

**Comment by nainar** [comment]
@alanorozco I have rebased on to the rewrite in #14276 so this should be much easier to read now.

File: extensions/amp-fx-collection/0.1/providers/fx-provider.js [hunk 13]
```diff
@@ -110,6 +118,19 @@ export class FxElement {
 
     /** @private {number} */
     this.factor_ = parseFloat(element.getAttribute('data-parallax-factor'));
+
```
Comment (alanorozco) [inline extensions/amp-fx-collection/0.1/providers/fx-provider.js]: What worries me about this style is that the specific FX are being scattered in more than one place.

For generic properties, like `easing` or `duration`, `data-duration` (without the preset name) should work fine. If they are to be invalid for certain presets, we can add validation rules.

File: extensions/amp-fx-collection/amp-fx-collection.md (hunk 10 above)
Comment (alanorozco) [reply extensions/amp-fx-collection/amp-fx-collection.md]: Awesome, thanks for updating the docs in the same PR!

I would add a note about fade-in being experimental.

File: extensions/amp-fx-collection/0.1/amp-fx-collection.js (hunk 2 above)
Comment (alanorozco) [reply extensions/amp-fx-collection/0.1/amp-fx-collection.js]: Can this be moved to the preset?

Comment (nainar) [reply extensions/amp-fx-collection/0.1/amp-fx-collection.js]: Done.

File: extensions/amp-fx-collection/0.1/providers/amp-fx-presets.js [hunk 14]
```diff
@@ -60,4 +60,39 @@ export const Presets = {
...
+      dev().assert(fxElement.adjustedViewportHeight_);
+      // Outside viewport
+      if (!entry.positionRect ||
+          entry.positionRect.top >
+            (1 - fxElement.margin_) * fxElement.adjustedViewportHeight_) {
+        return;
+      }
+
+      // If above the threshold of trigger-position
+      if (!fxElement.isMutateScheduled()) {
```
[the remark this answers is not in the document]
Comment (nainar) [reply extensions/amp-fx-collection/0.1/providers/amp-fx-presets.js]: I agree with the short circuiting here. However `FxElement` doesn't extend `BaseElement` so I don't think the mutateElement will work here, will it?

**Pull request merged**